File: src/main/cpp/BernoulliPowerSum.cc
#include <iostream>

using std::endl;

#include "BernoulliPowerSum.h"

BernoulliPowerSum :: BernoulliPowerSum()
                   : PowerSum() {
}

/* Bernoulli numbers are defined by the following recurrence relation:
 * B(0) = 1
 * B(m) = -(Binom((m + 1), 0)B(0) + Binom((m + 1), 1)B(1)
 *          + ... + Binom((m + 1), (m - 1))B(m - 1))
 * where Binom(i, j) is the binomial coefficient which evaluates to:
 * i!/{(i - j)!j!}
 */
vector<mpq_class> BernoulliPowerSum :: getCoefficients(long power) {
  vector<mpq_class> coeffs;

  if (power < 0) {
    return coeffs;
  }

  // Initialize B(0)
  coeffs.push_back(1);
  if (power > 0) {
    // Initialize B(1)
    coeffs.push_back(mpq_class(-1, 2));

    for (long i = 2; i <= power; i++) {
      if (i & 1) {
        // Odd coefficients above 1 are 0s
        coeffs.push_back(0);
      } else {
        coeffs.push_back(computeNextCoefficient(coeffs, i));
      }
    }
  }
  return coeffs;
}

void BernoulliPowerSum :: printSumFormula(long power, ostream &out) {
  if (power < 0) {
    return;
  }

  vector<mpq_class> coeffs = getCoefficients(power);
  mpz_class binom = 1;
  long binomN = power + 1;
  long binomR = 1;
  long p = power + 1;

  out << "{ ";
  for (size_t i = 0; i < coeffs.size(); i++) {
    if ((i & 1) == 0 || i == 1) {
      if (i != 0) {
        out << " + ";
      }
      if (coeffs[i] != 1) {
        out << '(' << coeffs[i] << ')';
      }
      if (binom != 1) {
        out << binom;
      }
      out << "(n + 1)";
      if (p != 1) {
        out << '^' << p;
      }
    }
    p--;
    binom = binom*binomN/binomR;
    binomN--;
    binomR++;
  }
  out << " }";
  if (power > 0) {
    out << '/' << (power + 1);
  }
  out << endl;
}

mpz_class BernoulliPowerSum :: computeSumWithTimeStat(long power, long n, 
                                                      vector<long> & stat) {
  struct timespec before;
  struct timespec after;

  mpq_class sum = 0;
  stat.clear();

  if (power < 0 || n < 0) {
    stat.push_back(0);
    stat.push_back(0);
    return sum.get_num();
  }

  clock_gettime(CLOCK_THREAD_CPUTIME_ID, &before);
  vector<mpq_class> coeffs = getCoefficients(power);
  clock_gettime(CLOCK_THREAD_CPUTIME_ID, &after);
  stat.push_back(computeCpuTime(before, after));

  clock_gettime(CLOCK_THREAD_CPUTIME_ID, &before);
  mpz_class pow = n + 1;
  long binomN = power + 1;
  long binomR = 1;
  mpz_class binom = binomN;

  // We compute the terms in reverse order to avoid an additional division
  // operation in the loop
  for (long i = coeffs.size() - 1; i >= 0; i--) {
    if ((i & 1) == 0 || i == 1) {
      sum += binom*coeffs[i]*pow;
    }
    pow *= (n + 1);
    binomN--;
    binomR++;
    binom = binom*binomN/binomR;
  }
  sum /= (power + 1);
  sum.canonicalize();
  clock_gettime(CLOCK_THREAD_CPUTIME_ID, &after);
  stat.push_back(computeCpuTime(before, after));
  return sum.get_num();
}

BernoulliPowerSum :: ~BernoulliPowerSum() {
}

mpq_class BernoulliPowerSum :: computeNextCoefficient(
                                       vector<mpq_class> & current, long m) {
  mpz_class binom = 1;
  mpq_class coeff = 0;

  for (long k = 0; k < m; k++) {
    if (((k & 1) == 0) || k == 1) {
      // Even coefficient or the first odd one
      coeff += current[k]*binom;
    }
    binom *= (m + 1 - k);
    binom /= (k + 1);
  }
  coeff = -coeff/binom;
  coeff.canonicalize();
  return coeff;
}

File: src/main/python/bernoulli_power_sum.py
from gmpy2 import mpz, mpq
from power_sum import PowerSum
import time

class BernoulliPowerSum(PowerSum):

  def __init__(self) -> None:
    return

  # Bernoulli numbers are defined by the following recurrence relation:
  # B(0) = 1
  # B(m) = -(Binom((m + 1), 0)B(0) + Binom((m + 1), 1)B(1)
  #          + ... + Binom((m + 1), (m - 1))B(m - 1))
  # where Binom(i, j) is the binomial coefficient which evaluates to:
  # i!/{(i - j)!j!}
  def get_coefficients(self, power: int) -> list[mpq]:
    coeffs: list[mpq] = []
    if power < 0:
      return coeffs

    i: int

    for i in range(0, power + 1):
     coeffs.append(0)

    # Initialize B(0)
    coeffs[0] = mpq(1)
    if power > 0:
      # Initialize B(1)
      coeffs[1] = mpq(-1, 2)

    for i in range(2, power + 1):
      if ((i & 1) == 1):
        # Odd coefficients above 1 are 0s
        coeffs[i] = mpq(0)
      else:
        coeffs[i] = self.__compute_next_coefficient(coeffs, i)

    return coeffs

  def print_sum_formula(self, power: int) -> None:
    if power < 0:
      return

    coeffs: list[mpq] = self.get_coefficients(power)
    binom: mpz = 1
    binomN: int = power + 1
    binomR: int = 1
    p: int = power + 1

    print('{ ', end='')
    for i in range(0, len(coeffs)):
      if ((i & 1) == 0 or i == 1):
        if (i != 0):
          print(' + ', end='')
        if (coeffs[i] != 1):
          print(f'({coeffs[i]})', end='')
        if (binom != 1):
          print(binom, end='')
        print('(n + 1)', end='')
        if (p != 1):
          print(f'^{p}', end='')
      p -= 1
      binom = binom*binomN//binomR
      binomN -= 1
      binomR += 1
    
    print(' }', end='')
    if (power > 0):
      print(f'/{(power + 1)}', end='')
    print()

  def compute_sum_with_time_stat(self, power: int, n: int,
                                 stat: list[int]) -> mpz:
    stat.clear()

    if (power < 0 or n < 0):
      stat.append(0)
      stat.append(0)
      result: mpz = 0
      return result

    before = time.clock_gettime_ns(time.CLOCK_THREAD_CPUTIME_ID)
    coeffs: list[mpq] = self.get_coefficients(power)
    after = time.clock_gettime_ns(time.CLOCK_THREAD_CPUTIME_ID)
    stat.append(after - before)

    before = time.clock_gettime_ns(time.CLOCK_THREAD_CPUTIME_ID)
    sum: mpq = 0
    p: mpz = n + 1
    binomN: int = power + 1
    binomR: int = 1
    binom: mpz = binomN

    # We compute the terms in reverse order to avoid an additional division
    # operation in the loop
    for i in range(len(coeffs) - 1, -1, -1):
      if ((i & 1) == 0 or i == 1):
        sum += binom*coeffs[i]*p
      p *= (n + 1)
      binomN -= 1
      binomR += 1
      binom = binom*binomN//binomR

    sum /= (power + 1)
    (numerator, denominator) = sum.as_integer_ratio()
    after = time.clock_gettime_ns(time.CLOCK_THREAD_CPUTIME_ID)
    stat.append(after - before)
    return numerator

  def __compute_next_coefficient(self, current: list[mpq], m: int) -> mpq:
    binom: mpz = 1
    coeff: mpq = 0
    k: int

    for k in range (0, m):
      if (((k & 1) == 0) or k == 1):
        # Even coefficient or the first odd one
        coeff += current[k]*binom
      binom *= (m + 1 - k)
      binom //= (k + 1)
    coeff = -coeff/binom
    return coeff
